---
title: "Preset_ST_Code"
author: "Jonathan"
date: "2024-02-19"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

Setup
```{r}
library(spdep)
library(rstan)
library(ggplot2)
library(tmap)
rstan_options(auto_write = TRUE)
options(mc.cores = parallel::detectCores())
library(SingleCellExperiment)
library(SummarizedExperiment)
library(MatrixGenerics)
library(matrixStats)
library(SpatialExperiment)
library(STexampleData)
library(ggspavis)
library(scuttle)
library(scater)
library(scran)
library(pheatmap)
library(nnSVG)
library(igraph)
library(BiocFileCache)
library(rtracklayer)
library(lobstr)
library(spatialLIBD)
```

Insert Code and Analysis
```{r}
spe <- 'Load data set here'
spe
dim(spe)
assayNames(spe)
head(rowData(spe))
head(colData(spe))
head(spatialCoords(spe))
imgData(spe)
dim(rowData(spe))
```

Quality Control

6.3 Plot data
```{r}
# plot spatial coordinates (spots)
plotSpots(spe)
```

6.4 Calculate QC metric
```{r}
# subset to keep only spots over tissue
spe <- spe[, colData(spe)$in_tissue == 1]
dim(spe)

# identify mitochondrial genes
is_mito <- grepl("(^MT-)|(^mt-)", rowData(spe)$gene_name)
table(is_mito)

rowData(spe)$gene_name[is_mito]

# calculate per-spot QC metrics and store in colData
spe <- addPerCellQC(spe, subsets = list(mito = is_mito))
head(colData(spe))
```


6.5.1 Library Size
```{r}
# histogram of library sizes
hist(colData(spe)$sum, breaks = 20)

# plot library size vs. number of cells per spot
plotQC(spe, type = "scatter", 
       metric_x = "cell_count", metric_y = "sum", 
       threshold_y = 600)

# select QC threshold for library size
qc_lib_size <- colData(spe)$sum < 600
table(qc_lib_size)

colData(spe)$qc_lib_size <- qc_lib_size

# check spatial pattern of discarded spots
plotQC(spe, type = "spots", 
       discard = "qc_lib_size")

#this is what happens when you set the threshhold too high
# check spatial pattern of discarded spots if threshold is too high
qc_lib_size_2000 <- colData(spe)$sum < 2000
colData(spe)$qc_lib_size_2000 <- qc_lib_size_2000
plotQC(spe, type = "spots", 
       discard = "qc_lib_size_2000")

# plot ground truth (manually annotated) layers
plotSpots(spe, annotate = "ground_truth", 
          palette = "libd_layer_colors")
```

6.5.2 Number of expressed features
```{r}
# histogram of numbers of expressed genes
hist(colData(spe)$detected, breaks = 20)

# plot number of expressed genes vs. number of cells per spot
plotQC(spe, type = "scatter", 
       metric_x = "cell_count", metric_y = "detected", 
       threshold_y = 400)

# select QC threshold for number of expressed genes
qc_detected <- colData(spe)$detected < 400
table(qc_detected)

colData(spe)$qc_detected <- qc_detected

# check spatial pattern of discarded spots
plotQC(spe, type = "spots", 
       discard = "qc_detected")

# check spatial pattern of discarded spots if threshold is too high
qc_detected_1000 <- colData(spe)$detected < 1000
colData(spe)$qc_detected_1000 <- qc_detected_1000
plotQC(spe, type = "spots", 
       discard = "qc_detected_1000")
```


6.5.3 Proportion of mitochondrial reads
```{r}
# histogram of mitochondrial read proportions
hist(colData(spe)$subsets_mito_percent, breaks = 20)

# plot mitochondrial read proportion vs. number of cells per spot
plotQC(spe, type = "scatter", 
       metric_x = "cell_count", metric_y = "subsets_mito_percent", 
       threshold_y = 28)

# select QC threshold for mitochondrial read proportion
qc_mito <- colData(spe)$subsets_mito_percent > 28
table(qc_mito)

colData(spe)$qc_mito <- qc_mito

# check spatial pattern of discarded spots
plotQC(spe, type = "spots", 
       discard = "qc_mito")

# check spatial pattern of discarded spots if threshold is too high
qc_mito_25 <- colData(spe)$subsets_mito_percent > 25
colData(spe)$qc_mito_25 <- qc_mito_25
plotQC(spe, type = "spots", 
       discard = "qc_mito_25")
```


6.5.4 Number of cells per spot
```{r}
# histogram of cell counts
hist(colData(spe)$cell_count, breaks = 20)

# distribution of cells per spot
tbl_cells_per_spot <- table(colData(spe)$cell_count)

# plot number of expressed genes vs. number of cells per spot
plotQC(spe, type = "scatter", 
       metric_x = "cell_count", metric_y = "detected", 
       threshold_x = 10)

# select QC threshold for number of cells per spot
qc_cell_count <- colData(spe)$cell_count > 10
table(qc_cell_count)

colData(spe)$qc_cell_count <- qc_cell_count

# check spatial pattern of discarded spots
plotQC(spe, type = "spots", 
       discard = "qc_cell_count")
```

6.5.5 Remove low-quality spots
```{r}
# number of discarded spots for each metric
apply(cbind(qc_lib_size, qc_detected, qc_mito, qc_cell_count), 2, sum)

# combined set of discarded spots
discard <- qc_lib_size | qc_detected | qc_mito | qc_cell_count
table(discard)

# store in object
colData(spe)$discard <- discard

# check spatial pattern of combined set of discarded spots
plotQC(spe, type = "spots", 
       discard = "discard")

# remove combined set of low-quality spots
spe <- spe[, !colData(spe)$discard]
dim(spe)
```


6.6 Zero-cell and single-cell spots
```{r}
# distribution of cells per spot
tbl_cells_per_spot[1:13]

# as proportions
prop_cells_per_spot <- round(tbl_cells_per_spot / sum(tbl_cells_per_spot), 2)
prop_cells_per_spot[1:13]
```