---
title: "Whole_Transcriptome"
author: "Jonathan"
date: "2024-03-13"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

library setup
```{r}
library(SpatialExperiment)
library(spdep)
library(scran)
library(scuttle)
library(dplyr)
library(ggspavis)
library(spdep)
library(scater)
library(pheatmap)
```

data setup
```{r}
dir <- file.path("/Users/jonathanhuang/Desktop/Boston_University/Research_Rotation/ST_Project/MSAB_Research_Rotation/Human_Breast_Cancer/Whole_Transcriptome")
samples <- file.path(dir, "outs")
wt<- read10xVisium(samples = samples)
spe_wt <- wt[, colData(wt)$in_tissue == 1]
is_mito <- grepl("(^MT-)|(^mt-)", rowData(spe_wt)$symbol)
spe_wt_f <- addPerCellQC(spe_wt, subsets = list(mito = is_mito))
```

Overview + Quality Control
```{r}
plotSpots(wt)
head(colData(wt))
head(rowData(wt))
head(spatialCoords(wt))

spe_wt <- wt[, colData(wt)$in_tissue == 1]
dim(spe_wt)

# idHBC# identify mitochondrial genes
is_mito <- grepl("(^MT-)|(^mt-)", rowData(spe_wt)$symbol)
table(is_mito)

rowData(spe_wt)$symbol[is_mito]

# calculate per-spot QC metrics and store in colData
spe_wt_f <- addPerCellQC(spe_wt, subsets = list(mito = is_mito))
head(colData(spe_wt_f))
```


Local Moran's I Implentation (sum of UMI)
```{r}
umi <- colData(spe_wt_f)$sum
id <- rownames(spatialCoords(spe_wt_f))
oid <- order(id)

nb_k6 <- knn2nb(knearneigh(spatialCoords(spe_wt_f), k=6))
resI_k6 <- localmoran(umi, nb2listw(nb_k6))
nb_k18 <- knn2nb(knearneigh(spatialCoords(spe_wt_f), k=18))
resI_k18 <- localmoran(umi, nb2listw(nb_k18))
nb_k36 <- knn2nb(knearneigh(spatialCoords(spe_wt_f), k=36))
resI_k36 <- localmoran(umi, nb2listw(nb_k36))

lmi_data_k6 <- printCoefmat(data.frame(resI_k6[oid,], row.names=rownames(spatialCoords(spe_wt_f))[oid]),
 check.names=FALSE)
lmi_data_k18 <- printCoefmat(data.frame(resI_k18[oid,], row.names=rownames(spatialCoords(spe_wt_f))[oid]),
 check.names=FALSE)
lmi_data_k36 <- printCoefmat(data.frame(resI_k36[oid,], row.names=rownames(spatialCoords(spe_wt_f))[oid]),
 check.names=FALSE)

colData(spe_wt_f)$Ii_sum_k6 <- lmi_data_k6$Ii
colData(spe_wt_f)$Ii_sum_k18 <- lmi_data_k18$Ii
colData(spe_wt_f)$Ii_sum_k36 <- lmi_data_k36$Ii

colData(spe_wt_f)$Ii_sum_k6 <- ifelse(colData(spe_wt_f)$Ii_sum_k6 < 0,0, colData(spe_wt_f)$Ii_sum_k6)
colData(spe_wt_f)$Ii_sum_k18 <- ifelse(colData(spe_wt_f)$Ii_sum_k18 < 0,0, colData(spe_wt_f)$Ii_sum_k18)
colData(spe_wt_f)$Ii_sum_k36 <- ifelse(colData(spe_wt_f)$Ii_sum_k36 < 0,0, colData(spe_wt_f)$Ii_sum_k36)

plotSpots(spe_wt_f, annotate = "Ii_sum_k6", pal = c("grey", "navy"),point_size = 1.2)+
  labs(title = "Sum of UMI with High AutoCorr (K = 6)")
plotSpots(spe_wt_f, annotate = "Ii_sum_k18", pal = c("grey", "navy"),point_size = 1.2)+
  labs(title = "Sum of UMI with High AutoCorr (K = 18)")
plotSpots(spe_wt_f, annotate = "Ii_sum_k36", pal = c("grey", "navy"),point_size = 1.2)+
  labs(title = "Sum of UMI with High AutoCorr (K = 36)")
```

Local Moran's I Implentation (detected features)
```{r}
umi <- colData(spe_wt_f)$detected
id <- rownames(spatialCoords(spe_wt_f))
oid <- order(id)

nb_k6 <- knn2nb(knearneigh(spatialCoords(spe_wt_f), k=6))
resI_k6 <- localmoran(umi, nb2listw(nb_k6))
nb_k18 <- knn2nb(knearneigh(spatialCoords(spe_wt_f), k=18))
resI_k18 <- localmoran(umi, nb2listw(nb_k18))
nb_k36 <- knn2nb(knearneigh(spatialCoords(spe_wt_f), k=36))
resI_k36 <- localmoran(umi, nb2listw(nb_k36))

lmi_data_k6 <- printCoefmat(data.frame(resI_k6[oid,], row.names=rownames(spatialCoords(spe_wt_f))[oid]),
 check.names=FALSE)
lmi_data_k18 <- printCoefmat(data.frame(resI_k18[oid,], row.names=rownames(spatialCoords(spe_wt_f))[oid]),
 check.names=FALSE)
lmi_data_k36 <- printCoefmat(data.frame(resI_k36[oid,], row.names=rownames(spatialCoords(spe_wt_f))[oid]),
 check.names=FALSE)

colData(spe_wt_f)$Ii_detected_k6 <- lmi_data_k6$Ii
colData(spe_wt_f)$Ii_detected_k18 <- lmi_data_k18$Ii
colData(spe_wt_f)$Ii_detected_k36 <- lmi_data_k36$Ii

colData(spe_wt_f)$Ii_detected_k6 <- ifelse(colData(spe_wt_f)$Ii_detected_k6 < 0,0, colData(spe_wt_f)$Ii_detected_k6)
colData(spe_wt_f)$Ii_detected_k18 <- ifelse(colData(spe_wt_f)$Ii_detected_k18 < 0,0, colData(spe_wt_f)$Ii_detected_k18)
colData(spe_wt_f)$Ii_detected_k36 <- ifelse(colData(spe_wt_f)$Ii_detected_k36 < 0,0, colData(spe_wt_f)$Ii_detected_k36)

plotSpots(spe_wt_f, annotate = "Ii_detected_k6", pal = c("grey", "navy"),point_size = 1.2)+
  labs(title = "Detected Features with High AutoCorr (K = 6)")
plotSpots(spe_wt_f, annotate = "Ii_detected_k18", pal = c("grey", "navy"),point_size = 1.2)+
  labs(title = "Detected Features with High AutoCorr (K = 18)")
plotSpots(spe_wt_f, annotate = "Ii_detected_k36", pal = c("grey", "navy"),point_size = 1.2)+
  labs(title = "Detected Features with High AutoCorr (K = 36)")
```

Local Moran's I Implentation (Mitochondrial Percentages)
```{r}
umi <- colData(spe_wt_f)$subsets_mito_percent
id <- rownames(spatialCoords(spe_wt_f))
oid <- order(id)

nb_k6 <- knn2nb(knearneigh(spatialCoords(spe_wt_f), k=6))
resI_k6 <- localmoran(umi, nb2listw(nb_k6))
nb_k18 <- knn2nb(knearneigh(spatialCoords(spe_wt_f), k=18))
resI_k18 <- localmoran(umi, nb2listw(nb_k18))
nb_k36 <- knn2nb(knearneigh(spatialCoords(spe_wt_f), k=36))
resI_k36 <- localmoran(umi, nb2listw(nb_k36))

lmi_data_k6 <- printCoefmat(data.frame(resI_k6[oid,], row.names=rownames(spatialCoords(spe_wt_f))[oid]),
 check.names=FALSE)
lmi_data_k18 <- printCoefmat(data.frame(resI_k18[oid,], row.names=rownames(spatialCoords(spe_wt_f))[oid]),
 check.names=FALSE)
lmi_data_k36 <- printCoefmat(data.frame(resI_k36[oid,], row.names=rownames(spatialCoords(spe_wt_f))[oid]),
 check.names=FALSE)

colData(spe_wt_f)$Ii_mit_k6 <- lmi_data_k6$Ii
colData(spe_wt_f)$Ii_mit_k18 <- lmi_data_k18$Ii
colData(spe_wt_f)$Ii_mit_k36 <- lmi_data_k36$Ii

colData(spe_wt_f)$Ii_mit_k6 <- ifelse(colData(spe_wt_f)$Ii_mit_k6 < 0,0, colData(spe_wt_f)$Ii_mit_k6)
colData(spe_wt_f)$Ii_mit_k18 <- ifelse(colData(spe_wt_f)$Ii_mit_k18 < 0,0, colData(spe_wt_f)$Ii_mit_k18)
colData(spe_wt_f)$Ii_mit_k36 <- ifelse(colData(spe_wt_f)$Ii_mit_k36 < 0,0, colData(spe_wt_f)$Ii_mit_k36)

plotSpots(spe_wt_f, annotate = "Ii_mit_k6", pal = c("grey", "navy"),point_size = 1.2)+
  labs(title = "Mitochondrial Percentages with High AutoCorr (K = 6)")
plotSpots(spe_wt_f, annotate = "Ii_mit_k18", pal = c("grey", "navy"),point_size = 1.2)+
  labs(title = "Mitochondrial Percentages with High AutoCorr (K = 18)")
plotSpots(spe_wt_f, annotate = "Ii_mit_k36", pal = c("grey", "navy"),point_size = 1.2)+
  labs(title = "Mitochondrial Percentages with High AutoCorr (K = 36)")
```

Nonspatial Clustering
```{r}
spe_2 <- logNormCounts(spe_wt_f)

dec <- modelGeneVar(spe_2)
top_hvgs <- getTopHVGs(dec, prop = 0.1)

# compute PCA
set.seed(123)
spe_2 <- runPCA(spe_2, subset_row = top_hvgs)
# compute UMAP on top 50 PCs
set.seed(123)
spe_2 <- runUMAP(spe_2, dimred = "PCA")
# update column names
colnames(reducedDim(spe_2, "UMAP")) <- paste0("UMAP", 1:2)

# graph-based clustering
set.seed(123)
k <- 50
g <- buildSNNGraph(spe_2, k = k, use.dimred = "PCA")
g_walk <- igraph::cluster_walktrap(g)
clus <- g_walk$membership
table(clus)

# store cluster labels in column 'label' in colData
colLabels(spe_2) <- factor(clus)

# plot clusters in spatial x-y coordinates
plotSpots(spe_2, annotate = "label", 
          pal = "libd_layer_colors",point_size = 1.2)

# plot clusters in PCA reduced dimensions
plotDimRed(spe_2, plot_type = "PCA", 
           annotate = "label", pal = "libd_layer_colors")

# plot clusters in UMAP reduced dimensions
plotDimRed(spe_2, plot_type = "UMAP", 
           annotate = "label", pal = "libd_layer_colors")

#Local Moran's I through Number of Detected Features
umi <- colData(spe_wt_f)$detected
id <- rownames(spatialCoords(spe_wt_f))
oid <- order(id)

nb_k6 <- knn2nb(knearneigh(spatialCoords(spe_wt_f), k=6))
resI_k6 <- localmoran(umi, nb2listw(nb_k6))

lmi_data <- printCoefmat(data.frame(resI_k6[oid,], row.names=rownames(spatialCoords(spe_2))[oid]),
 check.names=FALSE)
colData(spe_2)$Ii_detected <- lmi_data$Ii
plotSpots(spe_2, annotate = "Ii_detected",
          pal = c("grey", "navy"), point_size = 1.2) + 
labs(title = "K = 6")
```


Filtering for high autocorrlation (Ii > 0.5)
```{r}
spe_wt_filter <- spe_2[,colData(spe_2)$Ii_detected >0.5]
plotSpots(spe_wt_filter, annotate = "label", 
          pal = "libd_layer_colors",point_size = 1.2) + 
labs(title = "Detected Features with High AutoCorr (Li > 0.5 with K = 6)")

#Analzying the proportion of high Auto Corr with the clusters
total_corr <- length(colData(spe_wt_filter)$Ii_detected)

hc_c1 <- sum(colData(spe_wt_filter)$label == 1)
p_c1 <- hc_c1/total_corr*100

hc_c2 <- sum(colData(spe_wt_filter)$label == 2)
p_c2 <- hc_c2/total_corr*100

hc_c3 <- sum(colData(spe_wt_filter)$label == 3)
p_c3 <- hc_c3/total_corr*100

hc_c4 <- sum(colData(spe_wt_filter)$label == 4)
p_c4 <- hc_c4/total_corr*100

hc_c5 <- sum(colData(spe_wt_filter)$label == 5)
p_c5 <- hc_c5/total_corr*100

hc_c6 <- sum(colData(spe_wt_filter)$label == 6)
p_c6 <- hc_c6/total_corr*100

hc_c7 <- sum(colData(spe_wt_filter)$label == 7)
p_c7 <- hc_c7/total_corr*100

hist(as.numeric(colData(spe_wt_filter)$label), xlab = "Cluster Type", main = paste("Cluster Distribution (Li > 0.5)"), right = FALSE, breaks = c(1,2,3,4,5,6,7))
```

Filtering for high autocorrlation (Ii > 1.5)
```{r}
spe_wt_filter <- spe_2[,colData(spe_2)$Ii_detected >1.5]
plotSpots(spe_wt_filter, annotate = "label", 
          pal = "libd_layer_colors", point_size = 1.2) + 
labs(title = "Detected Features with High AutoCorr (Li > 1.5 with K = 6)")

#Analzying the proportion of high Auto Corr with the clusters
total_corr <- length(colData(spe_wt_filter)$Ii_detected)

hc_c1 <- sum(colData(spe_wt_filter)$label == 1)
p_c1 <- hc_c1/total_corr*100

hc_c2 <- sum(colData(spe_wt_filter)$label == 2)
p_c2 <- hc_c2/total_corr*100

hc_c3 <- sum(colData(spe_wt_filter)$label == 3)
p_c3 <- hc_c3/total_corr*100

hc_c4 <- sum(colData(spe_wt_filter)$label == 4)
p_c4 <- hc_c4/total_corr*100

hc_c5 <- sum(colData(spe_wt_filter)$label == 5)
p_c5 <- hc_c5/total_corr*100

hc_c6 <- sum(colData(spe_wt_filter)$label == 6)
p_c6 <- hc_c6/total_corr*100

hc_c7 <- sum(colData(spe_wt_filter)$label == 7)
p_c7 <- hc_c7/total_corr*100

hist(as.numeric(colData(spe_wt_filter)$label), xlab = "Cluster Type", main = paste("Cluster Distribution (Li > 1.5)"), right = FALSE, breaks = c(1,2,3,4,5,6,7))
```

Filtering for high autocorrlation (Ii > 2.5)
```{r}
spe_wt_filter <- spe_2[,colData(spe_2)$Ii_detected >2.5]
plotSpots(spe_wt_filter,  annotate = "label", 
          pal = "libd_layer_colors", point_size = 1.2) + 
labs(title = "Detected Features with High AutoCorr (Li > 2.5 with K = 6)")

#Analzying the proportion of high Auto Corr with the clusters
total_corr <- length(colData(spe_wt_filter)$Ii_detected)

hc_c1 <- sum(colData(spe_wt_filter)$label == 1)
p_c1 <- hc_c1/total_corr*100

hc_c2 <- sum(colData(spe_wt_filter)$label == 2)
p_c2 <- hc_c2/total_corr*100

hc_c3 <- sum(colData(spe_wt_filter)$label == 3)
p_c3 <- hc_c3/total_corr*100

hc_c4 <- sum(colData(spe_wt_filter)$label == 4)
p_c4 <- hc_c4/total_corr*100

hc_c5 <- sum(colData(spe_wt_filter)$label == 5)
p_c5 <- hc_c5/total_corr*100

hc_c6 <- sum(colData(spe_wt_filter)$label == 6)
p_c6 <- hc_c6/total_corr*100

hc_c7 <- sum(colData(spe_wt_filter)$label == 7)
p_c7 <- hc_c7/total_corr*100

hist(as.numeric(colData(spe_wt_filter)$label), xlab = "Cluster Type", main = paste("Cluster Distribution (Li > 2.5)"), right = FALSE, breaks = c(1,2,3,4,5,6,7))
```

Filtering for high autocorrlation (Ii > 3.5)
```{r}
spe_wt_filter <- spe_2[,colData(spe_2)$Ii_detected >3.5]
plotSpots(spe_wt_filter,  annotate = "label", 
          pal = "libd_layer_colors", point_size = 1.2) + 
labs(title = "Detected Features with High AutoCorr (Li > 3.5 with K = 6)")

#Analzying the proportion of high Auto Corr with the clusters
total_corr <- length(colData(spe_wt_filter)$Ii_detected)

hc_c1 <- sum(colData(spe_wt_filter)$label == 1)
p_c1 <- hc_c1/total_corr*100

hc_c2 <- sum(colData(spe_wt_filter)$label == 2)
p_c2 <- hc_c2/total_corr*100

hc_c3 <- sum(colData(spe_wt_filter)$label == 3)
p_c3 <- hc_c3/total_corr*100

hc_c4 <- sum(colData(spe_wt_filter)$label == 4)
p_c4 <- hc_c4/total_corr*100

hc_c5 <- sum(colData(spe_wt_filter)$label == 5)
p_c5 <- hc_c5/total_corr*100

hc_c6 <- sum(colData(spe_wt_filter)$label == 6)
p_c6 <- hc_c6/total_corr*100

hc_c7 <- sum(colData(spe_wt_filter)$label == 7)
p_c7 <- hc_c7/total_corr*100

hist(as.numeric(colData(spe_wt_filter)$label), xlab = "Cluster Type", main = paste("Cluster Distribution (Li > 3.5)"), right = FALSE, breaks = c(1,2,3,4,5,6,7))
```

Filtering for high autocorrlation (Ii > 4.5)
```{r}
spe_wt_filter_4 <- spe_2[,colData(spe_2)$Ii_detected >4.5]
plotSpots(spe_wt_filter_4,  annotate = "label", 
          pal = "libd_layer_colors", point_size = 1.2) + 
labs(title = "Detected Features with High AutoCorr (Li > 4.5 with K = 6)")

#Analzying the proportion of high Auto Corr with the clusters
total_corr <- length(colData(spe_wt_filter_4)$Ii_detected)

hc_c1 <- sum(colData(spe_wt_filter_4)$label == 1)
p_c1 <- hc_c1/total_corr*100

hc_c2 <- sum(colData(spe_wt_filter_4)$label == 2)
p_c2 <- hc_c2/total_corr*100

hc_c3 <- sum(colData(spe_wt_filter_4)$label == 3)
p_c3 <- hc_c3/total_corr*100

hc_c4 <- sum(colData(spe_wt_filter_4)$label == 4)
p_c4 <- hc_c4/total_corr*100

hc_c5 <- sum(colData(spe_wt_filter_4)$label == 5)
p_c5 <- hc_c5/total_corr*100

hc_c6 <- sum(colData(spe_wt_filter_4)$label == 6)
p_c6 <- hc_c6/total_corr*100

hc_c7 <- sum(colData(spe_wt_filter_4)$label == 7)
p_c7 <- hc_c7/total_corr*100

hist(as.numeric(colData(spe_wt_filter_4)$label), xlab = "Cluster Type", main = paste("Cluster Distribution (Li > 4.5)"), right = FALSE, breaks = c(1,2,3,4,5,6,7))
```

Filtering for high autocorrlation (Ii > 5.5)
```{r}
spe_wt_filter <- spe_2[,colData(spe_2)$Ii_detected >5.5]
plotSpots(spe_wt_filter,  annotate = "label", 
          pal = "libd_layer_colors", point_size = 1.2) + 
labs(title = "Detected Features with High AutoCorr (Li > 5.5 with K = 6)")

#Analzying the proportion of high Auto Corr with the clusters
total_corr <- length(colData(spe_wt_filter)$Ii_detected)

hc_c1 <- sum(colData(spe_wt_filter)$label == 1)
p_c1 <- hc_c1/total_corr*100

hc_c2 <- sum(colData(spe_wt_filter)$label == 2)
p_c2 <- hc_c2/total_corr*100

hc_c3 <- sum(colData(spe_wt_filter)$label == 3)
p_c3 <- hc_c3/total_corr*100

hc_c4 <- sum(colData(spe_wt_filter)$label == 4)
p_c4 <- hc_c4/total_corr*100

hc_c5 <- sum(colData(spe_wt_filter)$label == 5)
p_c5 <- hc_c5/total_corr*100

hc_c6 <- sum(colData(spe_wt_filter)$label == 6)
p_c6 <- hc_c6/total_corr*100

hc_c7 <- sum(colData(spe_wt_filter)$label == 7)
p_c7 <- hc_c7/total_corr*100

hist(as.numeric(colData(spe_wt_filter)$label), xlab = "Cluster Type", main = paste("Cluster Distribution (Li > 5.5)"), right = FALSE, breaks = c(1,2,3,4,5,6,7))
```

Filtering for high autocorrlation (Ii > 5.5)
```{r}
spe_wt_filter <- spe_2[,colData(spe_2)$Ii_detected >6.5]
plotSpots(spe_wt_filter,  annotate = "label", 
          pal = "libd_layer_colors", point_size = 1.2) + 
labs(title = "Detected Features with High AutoCorr (Li > 6.5 with K = 6)")

#Analzying the proportion of high Auto Corr with the clusters
total_corr <- length(colData(spe_wt_filter)$Ii_detected)

hc_c1 <- sum(colData(spe_wt_filter)$label == 1)
p_c1 <- hc_c1/total_corr*100

hc_c2 <- sum(colData(spe_wt_filter)$label == 2)
p_c2 <- hc_c2/total_corr*100

hc_c3 <- sum(colData(spe_wt_filter)$label == 3)
p_c3 <- hc_c3/total_corr*100

hc_c4 <- sum(colData(spe_wt_filter)$label == 4)
p_c4 <- hc_c4/total_corr*100

hc_c5 <- sum(colData(spe_wt_filter)$label == 5)
p_c5 <- hc_c5/total_corr*100

hc_c6 <- sum(colData(spe_wt_filter)$label == 6)
p_c6 <- hc_c6/total_corr*100

hc_c7 <- sum(colData(spe_wt_filter)$label == 7)
p_c7 <- hc_c7/total_corr*100

hist(as.numeric(colData(spe_wt_filter)$label), xlab = "Cluster Type", main = paste("Cluster Distribution (Li > 6.5)"), right = FALSE, breaks = c(1,2,3,4,5,6,7))
```


Differential Expression with region of interset
```{r}
x <- as.data.frame(spatialCoords(spe_wt_filter_4))
row <- 8000 #smaller than col
col <- 20000 #greater than row
test1_wt <- x[x$pxl_col_in_fullres < col & x$pxl_row_in_fullres < row,]

try1 <- spe_wt_filter_4[,which(row.names(colData(spe_wt_filter_4)) %in% row.names(test1_wt))]

plotSpots(try1, annotate = "label", 
          pal = "libd_layer_colors", point_size = 10)

test1_wt_v2 <- x[x$pxl_col_in_fullres > col & x$pxl_row_in_fullres < row,]
try1_v2 <- spe_wt_filter_4[,which(row.names(colData(spe_wt_filter_4)) %in% row.names(test1_wt_v2))]
plotSpots(try1_v2, annotate = "label", 
          pal = "libd_layer_colors", point_size = 1)



row_c3 <- 21000 #smaller than col
col_c3 <- 20000 #greater than row
test1_wt_v3 <- x[x$pxl_col_in_fullres > col_c3 & x$pxl_row_in_fullres > row_c3,]
try1_v3 <- spe_wt_filter_4[,which(row.names(colData(spe_wt_filter_4)) %in% row.names(test1_wt_v3))]
plotSpots(try1_v3, annotate = "label", 
          pal = "libd_layer_colors", point_size = 1)


spe_2 <- logNormCounts(spe_wt_f)

dec <- modelGeneVar(spe_2)
top_hvgs <- getTopHVGs(dec, prop = 0.1)

# compute PCA
set.seed(123)
spe_2 <- runPCA(spe_2, subset_row = top_hvgs)
# compute UMAP on top 50 PCs
set.seed(123)
spe_2 <- runUMAP(spe_2, dimred = "PCA")
# update column names
colnames(reducedDim(spe_2, "UMAP")) <- paste0("UMAP", 1:2)


spe_2_test <- spe_2
colLabels(spe_2_test) <- ifelse(row.names(colData(spe_2)) %in% row.names(test1_wt),1,4)
colLabels(spe_2_test) <- ifelse(row.names(colData(spe_2)) %in% row.names(test1_wt_v2),2,colLabels(spe_2_test))
colLabels(spe_2_test) <- ifelse(row.names(colData(spe_2)) %in% row.names(test1_wt_v3),3,colLabels(spe_2_test))
colLabels(spe_2_test)<-as.factor(colLabels(spe_2_test))

# plot clusters in spatial x-y coordinates
plotSpots(spe_2_test, annotate = "label", 
          pal = c("blue","red","green","grey") ,point_size = 1.2) +
labs(title = "Cluster of Interest (3 regions) vs. Others")

# plot clusters in PCA reduced dimensions
plotDimRed(spe_2_test, plot_type = "PCA", 
           annotate = "label", pal = c("blue","red","green","grey"))

# plot clusters in UMAP reduced dimensions
plotDimRed(spe_2_test, plot_type = "UMAP", 
           annotate = "label", pal = c("blue","red","green","grey"))

rownames(spe_2_test) <- rowData(spe_2)$symbol
markers_test <- findMarkers(spe_2_test, test = "binom", direction = "up")
interesting_test <- markers_test[[1]]

best_set <- interesting_test[interesting_test$Top <= 5, ]
logFCs <- getMarkerEffects(best_set)

pheatmap(logFCs, breaks = seq(-5, 5, length.out = 101), main = "Cluster of Interest (3 regions) vs. Others (top genes)")


markers_test_bot <- findMarkers(spe_2_test, test = "binom", direction = "down")
interesting_test_bot <- markers_test_bot[[1]]

best_set_bot <- interesting_test_bot[interesting_test_bot$Top <= 5, ]
logFCs_bot <- getMarkerEffects(best_set_bot)

pheatmap(logFCs_bot, breaks = seq(-5, 5, length.out = 101), main = "Cluster of Interest (3 regions) vs. Others (Bottom genes)")
```

Region 1 vs others
```{r}
colLabels(spe_2_test) <- ifelse(row.names(colData(spe_2)) %in% row.names(test1_wt),1,2)
rownames(spe_2_test) <- rowData(spe_2)$symbol

markers_c1 <- findMarkers(spe_2_test, test = "binom", direction = "up")

interesting_c1 <- markers_c1[[1]]
interesting_c1_upper <- interesting_c1[order(-interesting_c1$logFC.2), ]
sit_c1_top <- as.data.frame(head(interesting_c1_upper,20))
top_genes_c1 <- head(rownames(interesting_c1_upper),20)

interesting_c1_bottom <- interesting_c1[order(interesting_c1$logFC.2), ]
sit_c1_bot <- as.data.frame(head(interesting_c1_bottom,20))
bot_genes_c1 <- head(rownames(interesting_c1_bottom),20)
```


Region 2 vs others
```{r}
colLabels(spe_2_test) <- ifelse(row.names(colData(spe_2)) %in% row.names(test1_wt_v2),1,2)
rownames(spe_2_test) <- rowData(spe_2)$symbol

markers_c2 <- findMarkers(spe_2_test, test = "binom", direction = "up")

interesting_c2 <- markers_c2[[1]]
interesting_c2_upper <- interesting_c2[order(-interesting_c2$logFC.2), ]
sit_c2_top <- as.data.frame(head(interesting_c2_upper,20))
top_genes_c2 <- head(rownames(interesting_c2_upper),20)

interesting_c2_bottom <- interesting_c2[order(interesting_c2$logFC.2), ]
sit_c2_bot <- as.data.frame(head(interesting_c2_bottom,20))
bot_genes_c2 <- head(rownames(interesting_c2_bottom),20)
```

Region 2 vs others
```{r}
colLabels(spe_2_test) <- ifelse(row.names(colData(spe_2)) %in% row.names(test1_wt_v3),1,2)
rownames(spe_2_test) <- rowData(spe_2)$symbol

markers_c3 <- findMarkers(spe_2_test, test = "binom", direction = "up")

interesting_c3 <- markers_c3[[1]]
interesting_c3_upper <- interesting_c3[order(-interesting_c3$logFC.2), ]
sit_c3_top <- as.data.frame(head(interesting_c3_upper,20))
top_genes_c3 <- head(rownames(interesting_c3_upper),20)

interesting_c3_bottom <- interesting_c3[order(interesting_c3$logFC.2), ]
sit_c3_bot <- as.data.frame(head(interesting_c3_bottom,20))
bot_genes_c3 <- head(rownames(interesting_c3_bottom),20)
```